Closes #53 - invalidate all unreachable steps on forking

joefitter opened the pull request:
* Check for req.forkTaken on controller complete, if present then exec invalidation logic
* Get all reachable steps from path to invalidate, and path to validate, invalidate all in blacklist that aren't present in whitelist
* invalidation removes step and all step fields from sessionModel

joefitter commented:
Linked PR - https://github.com/UKHomeOffice/passports-form-controller/pull/41

joefitter commented:
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin this change fixes the current issue with forking by "invalidating" unreachable steps and fields when a fork is taken - if you go back and edit a field/step causing a fork condition to be evaluated - the losing path is wiped - this takes into account any potential shared steps and leaves these - logic is:

Difference of all reachable steps from losing fork and all reachable steps from winning fork.

gavboulton commented:
I haven't looked at these PRs yet but do we also need to cover the concept of branching? If forking is a two or more pathways, what happens when a user can go from a to c and some users may need to go via b1 and some via b2?

JoeChapman commented:
We should probably move the discussion to an issues so we can think through the logic and expectations clearly.

joefitter commented:
hey @gavboulton we cover all cases with forking logic - a fork can either add x steps to a journey, skip x steps or go down a completely different route - it all depends what you put as the target property

joefitter commented:
@gavboulton @JoeChapman @daniel-ac-martin @easternbloc anyone had a chance to review this yet? Linked PR https://github.com/UKHomeOffice/passports-form-controller/pull/41

JoeChapman commented:
This LGTM, @gavboulton do you see any problems with this?

gavboulton commented:
Looking.

gavboulton commented:
Can we have some tests please.

joefitter commented:
Ah - did write a load originally in hof-controllers where this logic lived - I must have forgotten to copy them over - I'll move them now

joefitter commented:
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin I have added tests, also changed the logic slightly so it doesn't require a change to form-controller

JoeChapman commented:
It's hard to read, but I get why. So it LGTM

joefitter commented:
I know - its quite complicated logic, basically get all reachable steps from the next step (after forking conditional executed), get all reachable steps from any route that isn't taken ((fork targets + next) - next step from forking conditional), and invalidate all steps that are reachable by routes not taken, but not reachable by route taken

joefitter commented:
@gavboulton is this ok to merge?

JoeChapman commented:
@joefitter @gavboulton I think the tests look good, @gavboulton are you ok to merge this?

gavboulton commented:
The tests describe what's actually being checked and not documenting the functionality. Could they be revised a bit please?

joefitter commented:
@gavboulton not sure I follow you - could you expand a bit please?

joefitter commented:
@gavboulton are you able to give an example of how you would like the tests to look?

gavboulton commented:
I'll take a look.

joefitter commented:
@gavboulton did you get a chance to look at this? Is it just the wording of the tests you would like changing or what is actually being tested?

gavboulton commented:
@joefitter It's the wording.

joefitter commented:
@gavboulton are you happy for this PR to go in?

gavboulton commented:
Not yet. The wording of the tests still don't describe features.

JoeChapman commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
+            checkProgress('/step1', controller, steps, '/');
+        });
+
+        describe('controller.getNextStep returns \'/fork1\'', function () {
```
revise to 'when getNextStep returns fork1`

JoeChapman commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
+                controller.emit('complete', req, res);
+            });
+
+            it('doesn\'t remove any steps from sessionModel', function () {
```
revise to 'it doesn't remove any steps' or 'no steps are removed'

JoeChapman commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,173 @@ describe('middleware/check-session', function () {
         });
     });
 
+    describe('step invalidation', function () {
```
Maybe revise to 'invalidates steps'

joefitter commented:
@gavboulton changed the wording as per @JoeChapman 's comments. Is this what you meant?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+            checkProgress('/step1', controller, steps, '/');
+        });
+
+        describe('when getNextStep returns \'/fork1\'', function () {
```
Do these tests describe a general forked journey?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+            });
+        });
+
+        describe('when getNextStep returns \'/step2\'', function () {
```
Is this when a journey no longer follows a fork?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+
+        });
+
+        describe('when getNextStep returns \'/step4\'', function () {
```
Is this when a step can be reached as part of normal and forked journeys?

joefitter commented:
Ah I think I get you @gavboulton

joefitter commented:
@gavboulton @easternbloc @JoeChapman @HughePaul @daniel-ac-martin can I get a couple of 👍's?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -1,7 +1,8 @@
+    helpers = require('../../lib/util/helpers'),
+    _ = require('underscore');
 
 describe('middleware/check-session', function () {
```
This is an old mistake but should be 'middleware/check-progress'. I'll raise it as a separate PR.

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
         });
     });
 
+    describe('invalidates steps', function () {
```
There are also fields and sometimes they aren't invalidated.

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+            });
+        });
+
+        describe('When result of forking skips a single step', function () {
```
The tests in this block are for steps and fields relating to a fork. Have you bypassed a previous fork here?

joefitter replied:
The fork is invalidated as it is unreachable with the result of getNextStep

joefitter replied:
This is a 3 'pronged' fork, any outcome will technically be 'forking' - next, or either fork target

gavboulton replied:
'when a step is skipped'?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+              controller.emit('complete', req, res);
+            });
+
+            it('invalidates fork1', function () {
```
invalidates forked journey?

joefitter replied:
there are 2 forked journeys - step4 and fork one. in this case only fork1 should be invalidated as step4 is still reachable

gavboulton replied:
I see.

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                req.sessionModel.get('steps').should.not.contain('/fork1');
+            });
+
+            it('invalidates fork-field-1 and fork-field-2', function () {
```
invalidates fields associated with a forked journey

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                expect(req.sessionModel.get('fork-field-2')).to.be.undefined;
+            });
+
+            it('doens\'t invalidate step2', function () {
```
doesn't invalidate the next step in the current journey?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                req.sessionModel.get('steps').should.contain('/step2');
+            });
+
+            it('doens\'t remove step-2-field-1 or step-2-field-2', function () {
```
doesn't unset fields associated with the next step in the current journey?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                controller.emit('complete', req, res);
+            });
+
+            it('invalidates step2, step3, and fork1', function () {
```
invalidates skipped steps?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                req.sessionModel.get('steps').should.not.contain.any('/step2', '/step3', '/fork1');
+            });
+
+            it('doesn\'t invalidate step1 or step4', function () {
```
doesn't invalidate steps in current journey? Or maybe new journey depending on how you look at it.

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -50,4 +51,164 @@ describe('middleware/check-session', function () {
+                req.sessionModel.get('steps').should.contain.all('/step1', '/step4');
+            });
+
+            it('invalidates all fields from /step2, /step3 and /fork1', function () {
```
invalidates all fields associated with skipped steps?

joefitter commented:
updated @gavboulton

joefitter commented:
goalposts have moved ;) - rebasing

joefitter commented:
@gavboulton how is this looking now?

gavboulton commented:
Looking.

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -49,4 +49,164 @@ describe('middleware/check-session', function () {
+                controller.emit('complete', req, res);
+            });
+
+            it('invalidates fork1', function () {
```
invalidates skipped step?

gavboulton commented on test/middleware/spec.check-progress.js:
```diff
@@ -49,4 +49,164 @@ describe('middleware/check-session', function () {
+                req.sessionModel.get('steps').should.not.contain('/fork1');
+            });
+
+            it('invalidates fields associated with fork1', function () {
```
invalidates fields associated with skipped step?

joefitter commented:
@gavboulton thanks, updated

gavboulton commented:
hmpo-form-wizard@4.4.0 published.
